File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/list/TransactionsModel.java
package com.androidcollider.easyfin.transactions.list;

import android.content.Context;
import android.util.Pair;

import androidx.core.content.ContextCompat;

import com.androidcollider.easyfin.R;
import com.androidcollider.easyfin.common.managers.format.date.DateFormatManager;
import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager;
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager;
import com.androidcollider.easyfin.common.models.Transaction;
import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.repository.Repository;
import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;

/**
 * @author Ihor Bilous
 */

class TransactionsModel implements TransactionsMVP.Model {

    private final Repository repository;
    private final DateFormatManager dateFormatManager;
    private final NumberFormatManager numberFormatManager;
    private final Context context;
    private final String[] curArray, curLangArray;


    TransactionsModel(Repository repository,
                      DateFormatManager dateFormatManager,
                      NumberFormatManager numberFormatManager,
                      ResourcesManager resourcesManager,
                      Context context) {
        this.repository = repository;
        this.dateFormatManager = dateFormatManager;
        this.numberFormatManager = numberFormatManager;
        this.context = context;
        curArray = resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY);
        curLangArray = resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY_LANG);
    }

    @Override
    public Flowable<Pair<List<TransactionViewModel>,
            Pair<List<TransactionCategory>, List<TransactionCategory>>>> getTransactionAndTransactionCategoriesLists() {
        return Flowable.combineLatest(
                repository.getAllTransactions()
                        .map(this::transformTransactionListToViewModelList),
                repository.getAllTransactionIncomeCategories(),
                repository.getAllTransactionExpenseCategories(),
                (transactionViewModels, transactionCategoryIncomeList, transactionCategoryExpenseList) ->
                        new Pair<>(
                                transactionViewModels,
                                new Pair<>(
                                        transactionCategoryIncomeList,
                                        transactionCategoryExpenseList
                                )
                        ));
    }

    @Override
    public Flowable<Transaction> getTransactionById(int id) {
        return repository.getAllTransactions()
                .flatMap(Flowable::fromIterable)
                .filter(transaction -> transaction.getId() == id);
    }

    @Override
    public Flowable<Boolean> deleteTransactionById(int id) {
        return getTransactionById(id)
                .flatMap(transaction ->
                        repository.deleteTransaction(
                                transaction.getIdAccount(),
                                transaction.getId(),
                                transaction.getAmount()
                        )
                );
    }

    private TransactionViewModel transformTransactionToViewModel(Transaction transaction) {
        TransactionViewModel model = new TransactionViewModel();

        model.setId(transaction.getId());
        model.setAccountName(transaction.getAccountName());
        model.setDate(dateFormatManager.longToDateString(transaction.getDate(), DateFormatManager.DAY_MONTH_YEAR_DOTS));

        String amount = numberFormatManager.doubleToStringFormatter(
                transaction.getAmount(),
                NumberFormatManager.FORMAT_1,
                NumberFormatManager.PRECISE_1
        );
        String curLang = null;

        for (int i = 0; i < curArray.length; i++) {
            if (transaction.getCurrency().equals(curArray[i])) {
                curLang = curLangArray[i];
                break;
            }
        }

        boolean isExpense = amount.contains("-");
        model.setExpense(isExpense);

        if (isExpense) {
            model.setAmount(String.format("- %1$s %2$s", amount.substring(1), curLang));
            model.setColorRes(ContextCompat.getColor(context, R.color.custom_red));
        } else {
            model.setAmount(String.format("+ %1$s %2$s", amount, curLang));
            model.setColorRes(ContextCompat.getColor(context, R.color.custom_green));
        }

        model.setCategory(transaction.getCategory());
        model.setAccountType(transaction.getAccountType());

        return model;
    }

    private List<TransactionViewModel> transformTransactionListToViewModelList(List<Transaction> transactionList) {
        return Stream.of(transactionList).map(this::transformTransactionToViewModel).collect(Collectors.toList());
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/list/TransactionsModel.kt
package com.androidcollider.easyfin.transactions.list

import android.content.Context
import androidx.core.content.ContextCompat
import com.androidcollider.easyfin.R
import com.androidcollider.easyfin.common.managers.format.date.DateFormatManager
import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager
import com.androidcollider.easyfin.common.models.Transaction
import com.androidcollider.easyfin.common.models.TransactionCategory
import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
internal class TransactionsModel(
    private val repository: Repository,
    private val dateFormatManager: DateFormatManager,
    private val numberFormatManager: NumberFormatManager,
    resourcesManager: ResourcesManager,
    private val context: Context
) : TransactionsMVP.Model {

    private val curArray: Array<String>
    private val curLangArray: Array<String>

    init {
        curArray = resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY)
        curLangArray =
            resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY_LANG)
    }

    override val transactionAndTransactionCategoriesLists:
            Single<Pair<List<TransactionViewModel>,
                    Pair<List<TransactionCategory>, List<TransactionCategory>>>>
        get() = Single.zip(
            repository.allTransactions!!.map { transactionList: List<Transaction> ->
                transformTransactionListToViewModelList(
                    transactionList
                )
            },
            repository.allTransactionIncomeCategories!!,
            repository.allTransactionExpenseCategories!!
        ) { transactionViewModels: List<TransactionViewModel>,
            transactionCategoryIncomeList: List<TransactionCategory>,
            transactionCategoryExpenseList: List<TransactionCategory> ->
            Pair(
                transactionViewModels,
                Pair(
                    transactionCategoryIncomeList,
                    transactionCategoryExpenseList
                )
            )
        }

    override fun getTransactionById(id: Int): Single<Transaction> {
        return repository.allTransactions!!.flatMap { source: List<Transaction> ->
            Single.just(source.first { transaction: Transaction -> transaction.id == id })
        }
    }

    override fun deleteTransactionById(id: Int): Single<Boolean> {
        return getTransactionById(id)
            .flatMap { transaction: Transaction ->
                repository.deleteTransaction(
                    transaction.idAccount,
                    transaction.id,
                    transaction.amount
                )
            }
    }

    private fun transformTransactionToViewModel(transaction: Transaction): TransactionViewModel {
        val model = TransactionViewModel()
        model.id = transaction.id
        model.accountName = transaction.accountName
        model.date = dateFormatManager.longToDateString(
            transaction.date,
            DateFormatManager.DAY_MONTH_YEAR_DOTS
        )
        val amount = numberFormatManager.doubleToStringFormatter(
            transaction.amount,
            NumberFormatManager.FORMAT_1,
            NumberFormatManager.PRECISE_1
        )
        var curLang: String? = null
        for (i in curArray.indices) {
            if (transaction.currency == curArray[i]) {
                curLang = curLangArray[i]
                break
            }
        }
        val isExpense = amount.contains("-")
        model.isExpense = isExpense
        if (isExpense) {
            model.amount = String.format("- %1\$s %2\$s", amount.substring(1), curLang)
            model.colorRes = ContextCompat.getColor(context, R.color.custom_red)
        } else {
            model.amount = String.format("+ %1\$s %2\$s", amount, curLang)
            model.colorRes = ContextCompat.getColor(context, R.color.custom_green)
        }
        model.category = transaction.category
        model.accountType = transaction.accountType
        return model
    }

    private fun transformTransactionListToViewModelList(transactionList: List<Transaction>)
            : List<TransactionViewModel> {
        return transactionList
            .map { transaction: Transaction -> transformTransactionToViewModel(transaction) }
    }
}